sshexec.1
.TH SSHEXEC 1 sshexec

.SH NAME
sshexec - run a command through ssh(1) with normal command syntax

.SH SYNOPSIS
.B sshexec
.RB [ {
.RI [\fBssh=\fP ssh-command ]
.RI [\fBdir=\fP directory ]
.RB [ cd= ( strict | lax )]
.RB [[\fIfd\fP]{ > , >> , >| , < , <> }[ & ] = \fIfile\fP]
.RB [ asis= \fIasis-marker\fP
.RB [ nasis= \fIasis-count\fP]]
.BR } ]
.RI [ ssh-option ]\ ...\,
.I destination
.I command
.RI [ argument ]\ ...\,

.SH DESCRIPTION
The
.B sshexec
utility is a wrapper for SSH that makes it easy to run commands
directly in the SSH command line.
.PP
.B sshexec
passes any argument after
.B }
to
.I ssh-command
.RB ( ssh
if not specified), but it
rewrites
.I command
and the
.IR argument s
to one argument that can be passed into
.BR ssh (1)
to describe each argument as a separate argument.
It may also rewrite
.I destination
to remove information that is not supported by
.BR ssh (1)
and insert extra arguments after it
(it may also add a
.B --
argument immediately before
.IR destination )
to cause the remote shell to change working directory to
.IR directory ,
if specified, and execute the provided
.I command
and
.IR argument s
as a regular command rather than as shell code joined
together by spaces.

.SH OPTIONS
.B sshexec
options may be placed at the very beginning enclosed with
the arguments
.B {
and
.BR } .
.B sshexec
options, if any, shall be placed in the same
.B {
.BR } -group.
Any other option will be passed as is to the
.BR ssh (1)
utility or
.IR ssh-command .
The
.B sshexec
utility has a built-in list of options recognised by the
.BR ssh (1)
utility and will not allow anything matching this list.
The
.B sshexec
utility does not allow mixing options and operands: no
option may be placed after
.IR destination ,
such options will be treated as the
.I command
or an
.IR argument .
.PP
The following
.B sshexec
options are supported:
.TP
.BI ssh= ssh-command
Instead of looking for
.B ssh
in
.IR PATH ,
the
.B sshexec
utility shall use
.IR ssh-command ,
which it will look for in
.I PATH
if it is only a file name (does not contain a slash
.RB ( / )).
.TP
.BI dir= directory
In the remote, change working directory to
.I directory
before executing
.IR command .
.TP
.B cd=strict
Fail without executing the
.I command
if it's not possible to set
.I directory
as the remote working directory.
.TP
.B cd=lax
Continue (but warn) executing the
.I command
even if it's not possible to set
.I directory
as the remote working directory.
.TP
.IB \fR[\fPfd\fP]\fP >= file
After changing working directory (assuming one is specified),
create or truncate the specified
.I file
and open it for writing, using file descriptor number
.IR fd .
(Default
.I fd
is 1 (standard output).)
.TP
.IB \fR[\fPfd\fP]\fP >>= file
After changing working directory (assuming one is specified),
create the specified
.I file
if it does not exist and open it for writing in
append-mode, using file descriptor number
.IR fd .
(Default
.I fd
is 1 (standard output).)
.TP
.IB \fR[\fPfd\fP]\fP >|= file
After changing working directory (assuming one is specified),
create the specified
.IR file ,
but fail if it already exists, and open it for writing,
using file descriptor number
.IR fd .
(Default
.I fd
is 1 (standard output).)
.TP
.IB \fR[\fPfd\fP]\fP <= file
After changing working directory (assuming one is specified),
open the specified
.IR file ,
for reading, using file descriptor number
.IR fd .
(Default
.I fd
is 0 (standard input).)
.TP
.IB \fR[\fPfd\fP]\fP <>= file
After changing working directory (assuming one is specified),
open the specified
.IR file ,
for reading and writing, creating it if it does not
already exist, using file descriptor number
.IR fd .
(Default
.I fd
is 0 (standard input).)
.TP
.IB \fR[\fPfd\fP]\fP \fR{\fP>\fP,\fP>>\fP,\fP>|\fP}\fP&= file
Duplicate the file descriptor
.I fd
giving the new file descriptor the number
.IR file .
(Default
.I fd
is 1 (standard output).)
.TP
.IB \fR[\fPfd\fP]\fP \fR{\fP>\fP,\fP>>\fP,\fP>|\fP}\fP&=-
Close the file descriptor
.IR fd .
(Default
.I fd
is 1 (standard output).)
.TP
.IB \fR[\fPfd\fP]\fP \fR{\fP<\fP,\fP<>\fP}\fP&= file
Duplicate the file descriptor
.I fd
giving the new file descriptor the number
.IR file .
(Default
.I fd
is 0 (standard input).)
.TP
.IB \fR[\fPfd\fP]\fP \fR{\fP<\fP,\fP<>\fP}\fP&=-
Close the file descriptor
.IR fd .
(Default
.I fd
is 0 (standard input).)
.TP
.BI asis= asis-marker
Any
.I argument
equal to
.I asis-marker
will be skipped over and instead the next argument
(regardless of whether it too is equal to
.IR asis-marker )
will be interpreted as raw shell code string that
shall be inserted without escaping.
.TP
.BI masis= asis-count
If specified,
.I asis-marker
shall only have its specified effect up to
.I asis-count
times.

.SH OPERANDS
The following operands are supported:
.TP
.I destination
The destination to connect and log into. It shall be either in
the form
.RI [ user\fP\fB@ ] hostname [\fB:\fP directory ]
or in the form
.BR ssh [ exec ] :// [\fIuser @ ]\fIhostname\fP[ : \fIport\fP][ / \fIdirectory\fP].

.I user
shall be the name of the remote user. If not specified,
the name of the local user running the utility will be used.

.I hostname
shall be the address to the remote machine.

.I port
shall be the port or service name for the port to
connect to on the remote machine.

.I directory
shall be the directory to change the remote working directory.
This is an alternative (with the exact same behaviour)
to the
.B dir
option and cannot be combined with it.
.TP
.IR command \ [ argument ]\ ...\,
Whereas the
.BR ssh (1)
utility would simply join the
.I command
and
.I argument
arguments with a space between each of them and pass the result
to the remote shell for execution, the
.B sshexec
utility forces the remote shell to treat each of them
as a separate argument and causes the shell to execute
them as a non-builtin command.

.I command
must not contain an equals sign
.RB ( = )
or be just a dash
.RB (\(dq - \(dq).

.SH STDIN
The
.B sshexec
utility itself does not use the standard input.

.SH INPUT FILES
None.

.SH ENVIRONMENT VARIABLES
The following environment variables affect the execution of
.BR sshexec :
.TP
.I PATH
Default. See to the Base Definitions volume of POSIX.1-2017, Section 8.3, Other Environment Variables.
This environment variable affects where the
.B sshexec
utility can find the
.BR ssh (1)
utility or
.IR ssh-command .
.TP
.I SSHEXEC_SSH
If set and non-empty, it overrides the default value of
.I ssh-command
from
.B ssh
to the value of the variable.
.TP
.I SSHEXEC_OPTS_NO_ARG
List of options that
.B sshexec
shall interpret as
.BR ssh (1)
options that do not have any argument.
(Default is
.BR 46AaCfGgKkMNnqsTtVvXxYy ,
meaning the options
.BR -4 ,
.BR -6 ,
.BR -A ,
.BR -a ,
.BR -C ,
.BR -f ,
.BR -G ,
.BR -g ,
.BR -K ,
.BR -k ,
.BR -M ,
.BR -N ,
.BR -n ,
.BR -q ,
.BR -s ,
.BR -T ,
.BR -t ,
.BR -V ,
.BR -v ,
.BR -X ,
.BR -x ,
.BR -Y ,
and
.BR -y .)
.TP
.I SSHEXEC_OPTS_ARG
List of options that
.B sshexec
shall interpret as
.BR ssh (1)
options that have an argument.
(Default is
.BR BbcDEeFIiJLlmOoPpQRSWw ,
meaning the options
.BR -B ,
.BR -b ,
.BR -c ,
.BR -D ,
.BR -E ,
.BR -e ,
.BR -F ,
.BR -I ,
.BR -i ,
.BR -J ,
.BR -L ,
.BR -l ,
.BR -m ,
.BR -O ,
.BR -o ,
.BR -P ,
.BR -p ,
.BR -Q ,
.BR -R ,
.BR -S ,
.BR -W ,
and
.BR -w .)
.TP
.I SSHEXEC_OPTS_OPT_ATTACHED_ARG
List of options that
.B sshexec
shall interpret as
.BR ssh (1)
options that have an argument only if
there are additional characters after
the option character in the same
command line argument. (Default is
the empty string, meaning no options.)
.TP
.I SSHEXEC_OPTS_OPT_ARG
List of options that
.B sshexec
shall interpret as
.BR ssh (1)
options that have an argument if there
are additional characters after
the option character in the same
command line argument or if argument is
followed directly by another argument
which does not start with a dash
.RB ( - ).
(Default is
the empty string, meaning no options.)
.TP
.I SSHEXEC_LONG_OPTS_NO_ARG
Space-separated list of long options that
.B sshexec
shall interpret as
.BR ssh (1)
options that do not have any argument
unless it is followed directly by an
equals sign
.RB ( = )
in the same command line argument.
Options that do not start with two dashes
.RB ( -- )
are silently ignored. (Default is the
empty string, meaning no options.)
.TP
.I SSHEXEC_LONG_OPTS_ARG
Space-separated list of long options that
.B sshexec
shall interpret as
.BR ssh (1)
options that have an argument that must
either be specified in the next command
line argument or after an
equals sign
.RB ( = )
the shall directly follow the option
string in the same command line argument.
Options that do not start with two dashes
.RB ( -- )
are silently ignored. (Default is the
empty string, meaning no options.)
.TP
.I SSHEXEC_LONG_OPTS_OPT_ARG
Space-separated list of long options that
.B sshexec
shall interpret as
.BR ssh (1)
options that have an argument if it is
the option string is is directly followed
by equals sign
.RB ( = )
in the same command line argument or if
argument is followed directly by another
argument which does not start with a dash
.RB ( - ).
Options that do not start with two dashes
.RB ( -- )
are silently ignored. (Default is the
empty string, meaning no options.)
.PP
Other environment variables may affect the execution of the
.BR ssh (1)
utility.

.SH ASYNCHRONOUS EVENTS
Default.

.SH STDOUT
The
.B sshexec
utility itself does not use the standard output.

.SH STDERR
The standard error is used for diagnostic messages in the
.B sshexec
utility itself.

.SH OUTPUT FILES
None.

.SH EXTENDED DESCRIPTION
None.

.SH EXIT STATUS
The
.B sshexec
utility exits with the exit status of the
.BR ssh (1)
utility or with 255 if an error occurred.

.SH CONSEQUENCES OF ERRORS
Default.

.SH APPLICATION USAGE
None.

.SH EXAMPLES
None.

.SH RATIONALE
The restrictions on
.I command
is in place to avoid unspecified behaviour.

.SH NOTES
None.

.SH BUGS
The remote shell must be sufficiently similar to
.BR sh (1posix).
Namely, it must support the
.B cd
builtin command and the commands
.B exec
and
.B printf
is expected by POSIX. Additionally, it must support
.BR \(dq$(\ )\(dq ,
.BR \(aq\ \(aq ,
and
.BR && ,
and argument separation with the SP character.
The remote shell must also not treat any alphanumeric
character, underscore
.RB ( _ )
or slash
.RB ( / )
as special characters.

.SH FUTURE DIRECTIONS
None.

.SH SEE ALSO
.BR ssh (1),
.BR sshcd (1)

.SH AUTHORS
Mattias Andrée
.RI < m@maandree.se >


sshexec.c
/* See LICENSE file for copyright and license details. */
#include <ctype.h>
#include <errno.h>
#include <limits.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>


/**
 * The name of the process
 */
static const char *argv0 = "sshexec";

/**
 * Whether the process is sshcd(1),
 * otherwise it's sshexec(1)
 */
static int is_sshcd = 0;


/**
 * Print an error message and exit
 *
 * @param  ...  Message format string and arguments
 */
#define exitf(...) (fprintf(stderr, __VA_ARGS__), exit(255))


/**
 * Print usage synopsis and exit
 */
static void
usage(void)
{
	exitf("usage: %s [{ [ssh=command]%s [cd=(strict|lax)]%s }] [ssh-option] ... destination%s\n",
	      argv0,
	      is_sshcd ? "" : " [dir=directory]",
	      is_sshcd ? "" : " [[fd]{>,>>,>|,<>}[&]=file] [asis=asis-marker [nasis=asis-count]]",
	      is_sshcd ? "" : " command [argument] ...");
}


/**
 * File descriptor redirection
 */
struct redirection {
	/**
	 * First part of the redirection string, shall not be escaped
	 */
	const char *asis;

	/**
	 * Second part of the redirection string, shall be escaped;
	 * or `NULL` if the entire string is contained in `.asis`
	 */
	const char *escape;
};


/**
 * Command being constructed for ssh(1)
 */
static char *command = NULL;

/**
 * The number of bytes allocated to `command`
 */
static size_t command_size = 0;

/**
 * The number of bytes written to `command`
 */
static size_t command_len = 0;


/**
 * The directory to use as the remote working directory
 */
static const char *dir = NULL;

/**
 * The ssh command to use
 */
static const char *ssh = NULL;

/**
 * Whether the command shall fail on failure to
 * set remote working directory
 */
static int strict_cd;

/**
 * The string used to mark the next argument
 * for unescaped inclusion in the command string
 */
static const char *asis = NULL;

/**
 * The number of times `asis` may be recognised
 */
static unsigned long long int nasis = ULLONG_MAX;

/**
 * File redirections to apply after directory switch
 */
static struct redirection *redirections = NULL;

/**
 * The number of elements in `redirections`
 */
static size_t nredirections = 0;

/**
 * The destination command line operand
 */
static char *destination;


/**
 * Add text to `command`
 *
 * @param  TEXT:const char *  The text to add to `command`
 * @param  N:size_t           The number of bytes to write
 */
#define build_command(TEXT, N)\
	do {\
		build_command_reserve(N);\
		memcpy(&command[command_len], (TEXT), (N));\
		command_len += (N);\
	} while (0)

/**
 * Add text to `command`
 *
 * @param  TEXT:const char *  The text to add to `command`
 */
#define build_command_asis(S)\
	build_command(S, strlen(S))

/**
 * NUL-byte terminate `command`
 */
#define finalise_command()\
	build_command("", 1)


/**
 * Allocate space for `command`
 * 
 * @param  n  The number of additional bytes (from what has currently
 *            been written, rather than currently allocated) `command`
 *            should have room for
 */
static void
build_command_reserve(size_t n)
{
	if (n <= command_size - command_len)
		return;

	if (n < 512)
		n = 512;
	if (command_len + n > SIZE_MAX)
		exitf("%s: could not allocate enough memory\n", argv0);
	command_size = command_len + n;
	command = realloc(command, command_size);
	if (!command)
		exitf("%s: could not allocate enough memory\n", argv0);
}


/**
 * Escape a string and add it to `command`
 *
 * @param  arg  The string to escape and add
 */
static void
build_command_escape(const char *arg)
{
#define IS_ALWAYS_SAFE(C)   (isalnum((C)) || (C) == '_' || (C) == '/')
#define IS_INITIAL_SAFE(C)  (isalpha((C)) || (C) == '_' || (C) == '/')

	size_t n = 0;
	size_t lfs = 0;

	/* Quote empty string */
	if (!*arg) {
		build_command_asis("''");
		return;
	}

	/* If the string only contains safe characters, add it without escaping */
	while (IS_ALWAYS_SAFE(arg[n]))
		n += 1;
	if (!arg[n]) {
		build_command(arg, n);
		return;
	}

	/* Escape string, using quoted printf(1) statement */
	build_command_asis("\"$(printf '");
	goto start; /* already have a count of safe initial characters, let's add them immediately */
	while (*arg) {
		/* Since process substitution removes at least one terminal
		 * LF, we must hold off on adding them */
		if (*arg == '\n') {
			lfs += 1;
			arg = &arg[1];
			continue;
		}

		/* Adding any held back LF */
		if (lfs) {
			build_command_reserve(lfs * 2U);
			for (; lfs; lfs--) {
				command[command_len++] = '\\';
				command[command_len++] = 'n';
			}
		}

		/* Character is unsafe, escape it */
		if (!IS_ALWAYS_SAFE(*arg)) {
			build_command_reserve(4);
			command[command_len++] = '\\';
			command[command_len] = (((unsigned char)*arg >> 6) & 7) + '0';
			command_len += (command[command_len] != '0');
			command[command_len] = (((unsigned char)*arg >> 3) & 7) + '0';
			command_len += (command[command_len] != '0');
			command[command_len++] = ((unsigned char)*arg & 7) + '0';
			arg = &arg[1];
		}

		/* Add any safe characters as is */
		n = 0;
		while (IS_INITIAL_SAFE(arg[n]) || arg[n] == '_' || arg[n] == '/')
			n += 1;
		if (n) {
			while (IS_ALWAYS_SAFE(arg[n]))
				n += 1;
		start:
			build_command(arg, n);
			arg = &arg[n];
		}
	}
	build_command_asis("\\n')\"");
	/* Add any held back terminal LF's */
	if (lfs) {
		build_command_reserve(lfs + 2U);
		command[command_len++] = '\'';
		memset(&command[command_len], '\n', lfs);
		command_len += lfs;
		command[command_len++] = '\'';
	}

#undef IS_ALWAYS_SAFE
#undef IS_INITIAL_SAFE
}


/**
 * Construct `command`
 *
 * @param  argv  The command to execute remotely and its arguments (`NULL` terminated)
 */
static void
construct_command(char *argv[])
{
	int next_asis = 0;
	size_t i;

	/* Change directory */
	if (dir) {
		build_command_asis("cd -- ");
		build_command_escape(dir);
		build_command_asis(strict_cd ? " && " : " ; ");
	}

	/* Execute command */
	if (is_sshcd) {
		build_command_asis("exec \"$SHELL\" -l");
		goto command_constructed;
	}
	if (asis)
		build_command_asis("( env --");
	else
		build_command_asis("exec env --");
	for (; *argv; argv++) {
		if (asis && nasis && !strcmp(*argv, asis)) {
			nasis -= 1U;
			next_asis = 1;
		} else {
			build_command_asis(" ");
			if (next_asis) {
				next_asis = 0;
				build_command_asis(*argv);
			} else {
				build_command_escape(*argv);
			}
		}
	}
	if (asis)
		build_command_asis(" )");

	/* Set redirections */
	for (i = 0; i < nredirections; i++) {
		build_command_asis(" ");
		build_command_asis(redirections[i].asis);
		if (redirections[i].escape) {
			build_command_asis(" ");
			build_command_escape(redirections[i].escape);
		}
	}

	/* NUL-terminate `command` */
command_constructed:
	finalise_command();
}


/**
 * Replace "sshexec://" prefix in `destination`
 * with "ssh://" and remove the directory
 * component and, if it present, store the
 * directory to `directory`
 */
static void
extract_directory_from_destination(void)
{
	char *dest_dir_delim;
	if (!is_sshcd && !strncmp(destination, "sshexec://", sizeof("sshexec://") - 1U)) {
		memmove(&destination[sizeof("ssh") - 1U],
		        &destination[sizeof("sshexec") - 1U],
		        strlen(&destination[sizeof("sshexec") - 1U]) + 1U);
		goto using_ssh_prefix;
	} else if (is_sshcd && !strncmp(destination, "sshcd://", sizeof("sshcd://") - 1U)) {
		memmove(&destination[sizeof("ssh") - 1U],
		        &destination[sizeof("sshcd") - 1U],
		        strlen(&destination[sizeof("sshcd") - 1U]) + 1U);
		goto using_ssh_prefix;
	} else if (!strncmp(destination, "ssh://", sizeof("ssh://") - 1U)) {
	using_ssh_prefix:
		dest_dir_delim = strchr(&destination[sizeof("ssh://")], '/');
	} else {
		dest_dir_delim = strchr(destination, ':');
	}
	if (dest_dir_delim) {
		if (dir)
			exitf("%s: directory specified both in 'dir' option and in destination operand\n", argv0);
		*dest_dir_delim++ = '\0';
		dir = dest_dir_delim;
	}
}


/**
 * Get an environment variable or a default value
 * 
 * @param   var  The name of the environment variable
 * @param   def  The default value to return if the environment variable is not set
 * @return       The value of the environment variable, or the default value if unset
 */
static const char *
get(const char *var, const char *def)
{
	const char *ret = getenv(var);
	return ret ? ret : def;
}


/**
 * Check if a string appears as a word in another string
 * 
 * @param   set     The string to search in; words are separated by spaces
 * @param   sought  The string to search for
 * @return          1 if the string is found, 0 otherwise
 */
#if defined(__GNUC__)
__attribute__((__pure__))
#endif
static int
contains(const char *set, const char *sought)
{
	size_t m, n = strlen(sought);
	const char *p, *q;
	for (p = set; (q = strchr(p, ' ')); p = &q[1]) {
		m = (size_t)(q - p);
		if (m == n && !strncmp(p, sought, n))
			return 1;
	}
	return !strcmp(p, sought);
}


/**
 * Read and parse the sshexec options
 *
 * @param   argv  The arguments from the command line after
 *                the zeroth argument
 * @return        `argv` offset to skip pass any sshexec options
 */
static char **
parse_sshexec_options(char *argv[])
{
	const char *cd = NULL;
	const char *nasis_str = NULL;
	char *p;

	/* Check that we have any arguments to parse, and that we
	 * have the "{" mark the beginning of sshexec options. */
	if (!*argv || strcmp(*argv, "{"))
		goto end_of_options;
	argv++;

#define STORE_OPT(VARP, OPT)\
	if (!strncmp(*argv, OPT"=", sizeof(OPT))) {\
		if (*(VARP) || !*(*(VARP) = &(*argv)[sizeof(OPT)]))\
			usage();\
		continue;\
	}

	/* Get options */
	for (; *argv && strcmp(*argv, "}"); argv++) {
		/* Options recognised in both sshexec(1) and sshcd(1) */
		STORE_OPT(&ssh, "ssh")
		STORE_OPT(&cd, "cd")

		/* The rest of the options only recognised in sshexec(1) */
		if (is_sshcd)
			usage();

		/* Normal options */
		STORE_OPT(&dir, "dir")
		STORE_OPT(&asis, "asis")
		STORE_OPT(&nasis_str, "nasis")

		/* File descriptor redirections */
		p = *argv;
		while (isdigit(*p))
			p++;
		if (p[0] == '>')
			p = &p[1 + (p[1] == '>' || p[1] == '|')];
		else if (p[0] == '<')
			p = &p[1 + (p[1] == '>')];
		else
			usage();
		if (p[p[0] == '&'] != '=')
			usage(); 
		redirections = realloc(redirections, (nredirections + 1U) * sizeof(*redirections));
		if (!redirections)
			exitf("%s: could not allocate enough memory\n", argv0);
		if (*p == '&') {
			p = &p[1];
			memmove(p, &p[1], strlen(&p[1]) + 1U);
			if (isdigit(*p)) {
				p++;
				while (isdigit(*p))
					p++;
			} else if (*p == '-') {
				p++;
			}
			if (*p)
				usage();
			redirections[nredirections].escape = NULL;
		} else {
			*p++ = '\0';
			redirections[nredirections].escape = p;
		}
		redirections[nredirections++].asis = *argv;
	}

	/* Check that we have the "}" argument that marks the end of sshexec options */
	if (!*argv)
		usage();
	argv++;

#undef STORE_OPT

end_of_options:
	/* Parse options and set defaults */

	if (!ssh) {
		ssh = get("SSHEXEC_SSH", "");
		if (!*ssh)
			ssh = "ssh";
	}

	if (!cd)
		strict_cd = !is_sshcd;
	else if (!strcmp(cd, "strict"))
		strict_cd = 1;
	else if (!strcmp(cd, "lax"))
		strict_cd = 0;
	else
		usage();

	if (nasis_str) {
		char *end;
		if (!asis || !isdigit(*nasis_str))
			usage();
		errno = 0;
		nasis = strtoull(nasis_str, &end, 10);
		if ((!nasis && errno) || *end)
			usage();
	}

	return argv;
}


/**
 * Read and parse the ssh(1) options
 *
 * @param   argv       The return value of `parse_sshexec_options`
 * @param   nopts_out  Output parameter for the number of option
 *                     arguments, this includes all arguments the
 *                     return value offsets `argv` except the "--"
 *                     (if there is one), so this includes both
 *                     the options themselves and their arguments,
 *                     and joined options countas one
 * @return             `argv` offset to skip pass any ssh(1) options,
 *                     and any "--" immediately after them
 */
static char **
parse_ssh_options(char *argv[], size_t *nopts_out)
{
	const char *unarged_opts;
	const char *arged_opts;
	const char *optatarged_opts;
	const char *optarged_opts;
	const char *unarged_longopts;
	const char *arged_longopts;
	const char *optarged_longopts;
	const char *arg;
	size_t nopts = 0;

	/* Get option syntax from the environment */
	unarged_opts = get("SSHEXEC_OPTS_NO_ARG", "46AaCfGgKkMNnqsTtVvXxYy");
	arged_opts = get("SSHEXEC_OPTS_ARG", "BbcDEeFIiJLlmOoPpQRSWw");
	optatarged_opts = get("SSHEXEC_OPTS_OPT_ATTACHED_ARG", "");
	optarged_opts = get("SSHEXEC_OPTS_OPT_ARG", "");
	unarged_longopts = get("SSHEXEC_LONG_OPTS_NO_ARG", "");
	arged_longopts = get("SSHEXEC_LONG_OPTS_ARG", "");
	optarged_longopts = get("SSHEXEC_LONG_OPTS_OPT_ARG", "");

	/* Count option arguments from the command line */
	while (*argv) {
		/* Break at the first non-option */
		if (!strcmp(*argv, "--")) {
			argv++;
			break;
		} else if ((*argv)[0] != '-' || !(*argv)[1]) {
			break;
		}

		arg = *argv++;
		nopts++;
		if (arg[1] == '-') {
			/* Long option */
			if (strchr(arg, '=')) {
				/* Option has attach argument */
			} else if (contains(unarged_longopts, arg)) {
				/* Option cannot have an argument */
			} else if (contains(arged_longopts, arg)) {
				/* Option has detached argument  */
				if (!*argv)
					exitf("%s: argument missing for option %s\n", argv0, arg);
				argv++;
				nopts++;
			} else if (contains(optarged_longopts, arg)) {
				/* Long option with either detached argument or no argument */
				if (*argv && **argv != '-') {
					/* The option has a detached argument */
					argv++;
					nopts++;
				}
			} else {
				exitf("%s: unrecognised option %s\n", argv0, arg);
			}
		} else {
			/* Short option */
			char opt;
			arg++;
			while ((opt = *arg++)) {
				if (strchr(unarged_opts, opt)) {
					/* Option cannot have an argument */
				} else if (strchr(arged_opts, opt)) {
					/* Option must have an argument */
					if (*arg) {
						/* Argument is attached to option */
						break;
					} else if (*argv) {
						/* Argument is detached from option */
						argv++;
						nopts++;
						break;
					} else {
						exitf("%s: argument missing for option -%c\n", argv0, opt);
					}
				} else if (strchr(optatarged_opts, opt)) {
					/* Option may have an attached argument, but it cannot have a detached argument */
					break;
				} else if (strchr(optarged_opts, opt)) {
					/* Option may have an attached or detached argument */
					if (*arg) {
						/* Argument is attached to option */
					} else {
						/* Either there is no argument, or it is detached from the option */
						if (*argv && **argv != '-') {
							/* Argument exist and is detached. We assume that if the next
							 * argument in the command line is the option's argument unless
							 * it starts with '-'. */
							argv++;
							nopts++;
							break;
						}
					}
				} else {
					exitf("%s: unrecognised option -%c\n", argv0, opt);
				}
			}
		}
	}

	*nopts_out = nopts;
	return argv;
}


int
main(int argc_unused, char *argv[])
{
	char **opts, *p;
	size_t nopts;
	const char **args;
	size_t i;

	(void) argc_unused;

	/* Identify used command name */
	if (*argv)
		argv0 = *argv++;
	p = strrchr(argv0, '/');
	is_sshcd = !strcmp(p ? &p[1] : argv0, "sshcd");

	/* Parse options */
	argv = parse_sshexec_options(argv);
	argv = parse_ssh_options(opts = argv, &nopts);

	/* Check operand count and separate out `destination` from the command and arguments */
	destination = *argv++;
	if (!destination || (is_sshcd ? !!*argv : !*argv))
		usage();

	/* Validate `destination` */
	if (!strcmp(destination, "-"))
		exitf("%s: the command argument must not be \"-\"\n", argv0);
	else if (strchr(destination, '='))
		exitf("%s: the command argument must not contain an \'=\'\n", argv0);

	/* Parse command line operands */
	extract_directory_from_destination();
	construct_command(argv);

	/* Construct arguments for ssh(1) and execute */
	i = 0;
	args = calloc(5U + (size_t)is_sshcd + nopts, sizeof(*args));
	if (!args)
		exitf("%s: could not allocate enough memory\n", argv0);
	args[i++] = ssh;
	if (is_sshcd) {
		const char *pty_alloc_flag = get("SSHCD_PTY_ALLOC_FLAG", "-t");
		if (*pty_alloc_flag)
			args[i++] = pty_alloc_flag;
	}
	memcpy(&args[i], opts, nopts * sizeof(*opts));
	i += nopts;
	args[i++] = "--";
	args[i++] = destination;
	args[i++] = command;
	args[i++] = NULL;
#if defined(__GNUC__)
# pragma GCC diagnostic ignored "-Wcast-qual"
#endif
	execvp(ssh, (char **)args);
	exitf("%s: failed to execute %s: %s\n", argv0, ssh, strerror(errno));
}
